assignment-3.cpp

```diff
@@ -0,0 +1,116 @@
+#include <iostream>
+#include <string>
+#include <vector>
+#include <queue>
+#include <set>
+#define f first
+#define s second
+using namespace std;
+
+struct state {
+    string word;
+    pair<int, int> last_letter;
+    vector<vector<bool> > visited;
+};
+
+/*Initializes an array of pairs that includes the different moves
+we can make from a position in the grid*/
+
+void initializemoves(pair<int, int> *moves) {
+    moves[0] = make_pair(1, 0);
+    moves[1] = make_pair(1, 1);
+    moves[2] = make_pair(-1, 0);
+    moves[3] = make_pair(-1, 1);
+    moves[4] = make_pair(-1, -1);
+    moves[5] = make_pair(0, 1);
+    moves[6] = make_pair(0, -1);
+    moves[7] = make_pair(1, -1);
+}
+
+/* Makes a state struct
+word: the word formed
+last_letter: the position in the grid of the last letter added to the word
+parents_visited: the vector including which chars have been visited so far*/
+
+state make_state(string word, pair<int, int> last_letter, vector<vector<bool> > parents_visited) {
+    state st;
+    st.word = word;
+    st.last_letter = last_letter;
+    st.visited.resize(parents_visited.size());
+    for (int i = 0; i < st.visited.size(); i++) {
+        st.visited[i].resize(parents_visited[i].size());
+        for (int j = 0; j < parents_visited[i].size(); j++) {
+            st.visited[i][j] = parents_visited[i][j];
+        }
+    }
+    st.visited[last_letter.f][last_letter.s] = true;
+    return st;
+}
+
+
+/*BFS from st with 
+ nodes = the different words that can be formed
+ and an edge exists between 2 nodes if
+ we can get one word from the other in the grid*/
+
+set<string> bfs(int r, int c, char **grid, pair<int, int> st) {
+    pair<int, int> moves[8];
+    initializemoves(moves);
+    set<string> found;
+    
+    queue<state> Q;
+    string k(1, grid[st.f][st.s]);
+    if (!isPrefix(k)) {
+        return found;
+    }
+    
+    vector<vector<bool> > a(r, vector<bool>(c));
```

Comment:
`a` is also not a very meaningful name... How about `visited` or `initial_visited`?
You can also inline this statement in `make_state` below:
`Q.push(make_state(k, st, vector<vector<bool> >(r, vector<bool>(c)));`

(but giving it a good name is better IMO)